File: frontend/lang3s/src/db/schemas/ontology.ts
import { TextAnnotationTable } from "@/db/schema";
import {
  boolean,
  customType,
  index,
  integer,
  jsonb,
  pgTable,
  serial,
  text,
  unique,
} from "drizzle-orm/pg-core";

////////////////////////////////////////////////////////////////////////////////
// Ontology Table
////////////////////////////////////////////////////////////////////////////////

const ltree = customType<{ data: string }>({
  dataType() {
    return "ltree";
  },
});

export const OntologyTable = pgTable(
  "ontology",
  {
    id: serial("id").primaryKey(),
    name: text("name").notNull().unique(),
    parentId: integer("parent_id"),
    description: text("description"),
    isAttribute: boolean("is_attribute").default(false).notNull(),
    path: ltree("path").notNull(), // ltree type (custom)
    properties: jsonb("properties").$type<Record<string, any>>().default({}),
  },
  (table) => [
    index("idx_ontology_parent_id").on(table.parentId),
    index("idx_ontology_name").on(table.name),
    index("idx_ontology_path_gist").using("GIST", table.path),
  ],
);

////////////////////////////////////////////////////////////////////////////////
// Annotation Value to Ontology Table
////////////////////////////////////////////////////////////////////////////////

export const AnnotationToOntology = pgTable(
  "annotation_to_ontology",
  {
    id: serial("id").primaryKey(),
    ontologyId: integer("ontology_id")
      .notNull()
      .references(() => OntologyTable.id, { onDelete: "cascade" }),
    annotation: text("annotation_type_value").notNull(),
  },
  (table) => [
    unique("ato_ont_type_idx").on(table.ontologyId, table.annotation),
    index("ato_ont_id").on(table.ontologyId),
    index("ato_ont_type").on(table.annotation),
  ],
);

File: backend/packages/data/src/lang3s/data/models/ontology.py
from __future__ import annotations

from typing import TYPE_CHECKING, Any, Optional

from sqlalchemy import ForeignKey, Index, Integer, String, Text
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.ext.mutable import MutableDict
from sqlalchemy.orm import Mapped, mapped_column, relationship
from sqlalchemy_utils import Ltree, LtreeType

if TYPE_CHECKING:
    from .annotation_ontology_mapping import AnnotationOntologyMapping

from . import Base


class Ontology(Base):
    __tablename__ = "ontology"

    id: Mapped[int] = mapped_column(
        "id",
        Integer,
        primary_key=True,
        autoincrement=True,
        index=True,
    )
    name: Mapped[str] = mapped_column(
        "name",
        String,
        nullable=False,
        unique=True,
    )
    parent_id: Mapped[Optional[int]] = mapped_column(
        "parent_id",
        ForeignKey("ontology.id", ondelete="CASCADE"),
        nullable=True,
        index=True,
    )
    description: Mapped[Optional[str]] = mapped_column(
        "description",
        Text,
        nullable=True,
    )
    color: Mapped[str] = mapped_column(
        "color",
        String,
        nullable=False,
    )
    path: Mapped[Ltree] = mapped_column(
        "path",
        LtreeType,
        nullable=False,
    )
    properties: Mapped[dict[str, Any]] = mapped_column(
        "properties",
        MutableDict.as_mutable(JSONB),
        nullable=False,
        server_default="{}",
    )

    mappings: Mapped[list[AnnotationOntologyMapping]] = relationship(
        "AnnotationOntologyMapping",
        back_populates="ontology_item",
        lazy="selectin",
        cascade="all, delete-orphan",
    )

    parent: Mapped[Optional[Ontology]] = relationship(
        "Ontology",
        remote_side="Ontology.id",
        back_populates="children",
    )

    children: Mapped[list[Ontology]] = relationship(
        "Ontology",
        back_populates="parent",
        cascade="all, delete-orphan",
        lazy="selectin",
    )

    __table_args__ = (
        Index(
            "idx_ontology_properties",
            "properties",
            postgresql_using="GIN",
            postgresql_ops={"metadata": "jsonb_path_ops"},
        ),
        Index(
            "idx_ontology_path",
            "path",
            postgresql_using="GIST",
            postgresql_ops={"metadata": "jsonb_path_ops"},
        ),
    )

    def __repr__(self) -> str:
        return f"<%s id={self.id}>" % self.__class__.__name__
